### PR TITLE
Add energy_threshold parameter

### LINKED ISSUE #48: Add target energy based termination criterion

### PR DESCRIPTION
**Pull request by hhtong**
Closes #48

### CODE REVIEW

File: tabu/src/tabu_search.cc [hunk 1]
```diff
@@ -90,7 +92,8 @@ void TabuSearch::multiStartTabuSearch(long long timeLimitInMilliSecs,
     vector<vector<double>> C(bqp.nVars, vector<double>(bqp.nVars));
 
     for (long iter = 0; iter < numRestarts; iter++) {
-        if (useTimeLimit && (realtime_clock() - startTime) > timeLimitInMilliSecs) {
+        if ((bestSolutionQuality <= energyThreshold) ||
```
Comment (randomir) [inline tabu/src/tabu_search.cc:95]: Would it make sense to add this condition to the inner loop, in `simpleTabuSearch`? It probably would help break out earlier on huge problems? Not sure it's worth the mess. I'd at least document where the test for energy is done.

File: tabu/sampler.py [hunk 2]
```diff
@@ -105,6 +106,9 @@ def sample(self, bqm, initial_states=None, initial_states_generator='random',
             num_restarts (int, optional, default=1,000,000):
                 Number of tabu search restarts per read.
 
+            energy_threshold (float, optional):
+                Terminate when this energy threshold is surpassed.
```
Comment (randomir) [inline tabu/sampler.py:110]: I'd say something like ~ when energy _lower_ than `energy_threshold` is found ...

**Review by randomir — APPROVED** [review]
LGTM

File: tabu/tabu_search.pyx [hunk 3]
```diff
@@ -26,8 +27,16 @@ cdef class TabuSearch:
...
+    def __cinit__(self, 
+                  object Q, 
+                  object initSol, 
+                  int tenure, 
+                  int timeout, 
+                  int numRestarts, 
+                  object seed=None, 
+                  object energyThreshold=None) :
         cdef unsigned int _seed = time(NULL) if seed is None else seed
+        cdef double _energyThreshold = -numeric_limits[double].max() if energyThreshold is None else energyThreshold
```
Comment (randomir) [inline tabu/tabu_search.pyx:39]: How about using `-inf` instead? Functionally is the same, but conceptually nicer.

Comment (hhtong) [reply tabu/tabu_search.pyx]: Done!

**Review by randomir — APPROVED** [review]

**Pull request merged**